### PR TITLE
Draft a write-up of parallel eNKI preprocessing

### LINKED ISSUE #472: How to `datalad run` in parallel
Having a section on how to parallelize `datalad run` calls on a cluster would be a nice addition to the handbook.

I'm currently trying to run multiple `datalad run` calls in parallel (with `--explicit` set) and getting problems with git-locks.. I have a few ideas how I could avoid those in my specific case, but I assume that this would be relevant to many people using some form of HPC.

I think the section should highlight what the potential problems are (git-locks / multiple processes trying to commit at the same time) and how to avoid them. The section on [big analyses](http://handbook.datalad.org/en/latest/beyond_basics/101-161-biganalyses.html) is related/a good template for this (e.g. it mentions the `--explicit` flag), but it does not address locking issues. 

A gist or brief outline of how to code this up would be a great start for being pointed in the right direction. 

I assume that there are "intended" programming patterns and more of "anti-patterns". Having a short do's/dont's would be also an option.

### PR DESCRIPTION
**Pull request by adswa**
A write up of parallel preprocessing of large datasets on computational clusters with containerized pipelines and `datalad containers-run`. This fixes #472.

Ping @mih.

Potential TODO:
- re-computation with re-run

### CODE REVIEW

**Comment by adswa** [comment]
Removed WIP label despite some yet missing pieces (see TODOs in main comment), but it is good to go for a rough review, @mih.

**Comment by mih** [comment]
Is there a build of this PR somewhere on RTD?

**Comment by adswa** [comment]
huh, good question... There should be, but I can't find one. I'll investigate

**Comment by adswa** [comment]
No clue why the readthedocs test build isn't triggered. it works for other PRs, and I have just resynced the webhooks...

**Comment by adswa** [comment]
the f***... PR build also showed up here now... not sure how, but its fixed!

**Comment by adswa** [comment]
@mih I have finished the draft on eNKI processing. Feedback is welcome.

File: docs/beyond_basics/101-170-dataladrun.rst [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+Given that the subdatasets each contain >1000 branches, and that each ``merge`` would lead to a commit, in order to not inflate the history of the dataset with hundreds of merge commits, two `Octopus merges <https://git-scm.com/docs/git-merge#Documentation/git-merge.txt-octopus>`_ were done - one in each subdataset (``fmriprep`` and ``freesurfer``).
+
+.. findoutmore:: What is an octopus merge?
+
+   Usually a commit that arises from a merge has two *parent* commits: The *first parent* is the branch the merge is being performed from, in the example above, ``master``. The *second parent* is the branch that was merged into the first.
+
+
+   However, ``git merge`` is capable of merging more than two branches simultaneously if more than a single branch name is given to the command.
+   The resulting merge commit has as many parent as were involved in the merge.
+   If a commit has more than two parents, if is affectionately called an "Octopus" merge.
```
Comment (mih) [inline docs/beyond_basics/101-170-dataladrun.rst:400]: s/if/it

File: docs/beyond_basics/101-170-dataladrun.rst [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    Fast-forwarding to: job-107890.0
+    Trying simple merge with job-107890.1
+    Simple merge did not work, trying automatic merge.
+    ERROR: logs/CITATION.md: Not merging symbolic link changes.
+    fatal: merge program failed
+    Automated merge did not work.
+    Should not be doing an octopus.
+    Merge with strategy octopus failed.
+
+This merge conflict arose in the ``fmriprep`` subdataset an originated from the fact that each job generated a ``CITATION.md`` file with minimal individual changes.
```
Comment (mih) [inline docs/beyond_basics/101-170-dataladrun.rst:430]: Personally, I would mention that octopus cannot have conflicts, but that is also often no problem for per-sample processing.

That it doesn't work out of the box with fmriprep is a footnote in this context, and I would move the mention and explanation entirely into a findoutmore box.

**Review by mih — CHANGES_REQUESTED** [review]
This look great as-is, but I nevertheless left a few remarks to tweak the message here and there, and highlight a few things that I personally consider also/more relevant.

Thx!

File: docs/beyond_basics/101-170-dataladrun.rst [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+""""""""""""""""""""""
+
+An analysis dataset consists of the following components:
+
+- input data as a subdataset
+- ``pipelines`` container dataset as a subdataset
+- subdatasets to hold the results
+
+Following the benchmarks and tips in the chapter :ref:`chapter_gobig`, the amount of files produced by fMRIprep on 1300 subjects requires two datasets to hold them.
+In this particular computation, following the naming scheme and structure of fMRIpreps output directories, one subdataset is created for the `freesurfer <https://surfer.nmr.mgh.harvard.edu/>`_ results of fMRIprep in a subdataset called ``freesurfer``, and one for the minimally preprocessed input data in a subdataset called ``fmriprep``.
```
[the remark this answers is not in the document]
Comment (adswa) [reply docs/beyond_basics/101-170-dataladrun.rst:155]: ok

[the remark this answers is not in the document]
Comment (adswa) [reply docs/beyond_basics/101-170-dataladrun.rst:155]: I read it again, its outlined in line 62: 

> The preprocessing results will encompass several TB and about half a million files, and will therefore need to be split into two result datasets.

I'll use it below as well.

File: docs/beyond_basics/101-170-dataladrun.rst [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+   $ datalad get -n -r -R1 .
+
+   # make git-annex disregard the clones - they are meant to be thrown away
+   $ git submodule foreach --recursive git annex dead here
+
+   # checkout unique branches (names derived from job IDs) in both subdatasets
+   # to enable pushing the results without interference from other jobs
+   $ git -C fmriprep checkout -b "job-$JOBID"
+   $ git -C freesurfer checkout -b "job-$JOBID"
+
```
[the remark this answers is not in the document]
Comment (adswa) [reply docs/beyond_basics/101-170-dataladrun.rst:206]: but checking out ID-specific branches would be necessary regardless of dataset. Without the subdatasets, the superdataset would need, them, wouldn't it?

[the remark this answers is not in the document]
Comment (mih) [reply docs/beyond_basics/101-170-dataladrun.rst:206]: Yes, indeed.

File: docs/beyond_basics/101-170-dataladrun.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mih) [reply docs/beyond_basics/101-170-dataladrun.rst:155]: Ah, ok. I think I read the start on readthedocs and it was an older version.

**Pull request closed without merging**